### PR TITLE
crimson/seastore: add perfcounters in seastore

### PR DESCRIPTION
**Pull request by liu-chunmei**
Signed-off-by: chunmei-liu <chunmei.liu@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by liu-chunmei** [comment]
@athanatos please review. will create another PR for perfcounters dump.

**Comment by athanatos** [comment]
jenkins retest this please

**Comment by athanatos** [comment]
Also, there's a conflict.

**Comment by liu-chunmei** [comment]
@athanatos rebased, seems all checks has passed. please review.

File: src/crimson/os/seastore/segment_cleaner.h [hunk 1]
```diff
@@ -365,7 +366,7 @@ class SegmentCleaner : public JournalSegmentProvider {
   std::optional<seastar::promise<>> blocked_io_wake;
 
 public:
-  SegmentCleaner(config_t config, bool detailed = false)
+  SegmentCleaner(config_t config, PerfCounters *perf, bool detailed = false)
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/os/seastore/segment_cleaner.h:369]: @ronen-fr any suggestion to pass perfcounters to each part such as transaction_manager, segment_manager, etc.  perfcoutners will be used in  member function of each class.

File: src/crimson/os/seastore/seastore_perf_counters.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    ss_plb.add_u64_counter(
+      ss_sm_bytes_written, "sm_bytes_written",
+      "the number of bytes written by segmentmanager");
+    ss_plb.add_u64_counter(
+      ss_sm_segments_released, "sm_segments_released",
+      "the number of segments released by segmentmanager");
+    ss_plb.add_u64_counter(
+      ss_sm_bytes_released, "sm_bytes_released",
+      "the number of bytes released by segmentmanager");
+    ss_perf = ss_plb.create_perf_counters();
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore_perf_counters.h:56]: @ronen-fr can you explain it in detail? I am not very understand it.

File: src/crimson/os/seastore/seastore_perf_counters.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#pragma once
+
+#include "crimson/common/perf_counters_collection.h"
+
+namespace crimson::os::seastore {
+
+using PerfCountersRef = std::unique_ptr<PerfCounters>;
+enum {
+  seastore_perfcounters_first = 300000,
+  ss_tm_extents_mutated_total,
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore_perf_counters.h:12]: the classic osd define it as this,  just follow the other part usage model.

File: src/crimson/os/seastore/seastore_perf_counters.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// vim: ts=8 sw=2 smarttab
+#pragma once
+
+#include "crimson/common/perf_counters_collection.h"
+
+namespace crimson::os::seastore {
+
+using PerfCountersRef = std::unique_ptr<PerfCounters>;
+enum {
+  seastore_perfcounters_first = 300000,
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore_perf_counters.h]: crimson-osd use l_osd_first = 10000, rs_first = 20000, so I want to use 30000 here ( wrong type one more 0. change it to 30000 now). we can make it depend on last one enum, such as make rs_fist = l_osd_last, and seastore_perfcounters_first = rs_last, if it has to.

File: src/crimson/os/seastore/cache.cc [hunk 5]
```diff
@@ -21,8 +21,9 @@ namespace {
 
 namespace crimson::os::seastore {
 
-Cache::Cache(SegmentManager &segment_manager) :
-  segment_manager(segment_manager) {}
+Cache::Cache(SegmentManager &segment_manager, PerfCounters *perf) :
+  segment_manager(segment_manager),
+  c_perf(perf) {}
```
Comment (athanatos) [inline src/crimson/os/seastore/cache.cc:26]: c_perf appears unused?

File: src/crimson/os/seastore/journal.cc [hunk 6]
```diff
@@ -164,7 +164,7 @@ bool Journal::validate_metadata(const bufferlist &bl)
     ceph::encoded_sizeof_bounded<record_header_t>(),
     -1);
   ceph_le32 recorded_crc_le;
-  ::decode(recorded_crc_le, bliter);
+  decode(recorded_crc_le, bliter);
```
Comment (athanatos) [inline src/crimson/os/seastore/journal.cc:167]: This should be in its own commit.

File: src/crimson/os/seastore/seastore_perf_counters.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    ss_perf = ss_plb.create_perf_counters();
+  }
+
+public:
+  PerfCounters *ss_perf = nullptr;
+  PerfService(){
+    build_seastore_perf();
+  }
+  ~PerfService() {
+    delete ss_perf;
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_perf_counters.h:65]: Use a unique_ptr for ss_perf.

File: src/crimson/os/seastore/seastore_perf_counters.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  ss_tm_extents_allocated_total,
+  ss_tm_extents_allocated_bytes,
+  ss_sm_bytes_written,
+  ss_sm_segments_released,
+  ss_sm_bytes_released,
+  seastore_perfcounters_last,
+};
+
+class PerfService {
+  void build_seastore_perf() {
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore_perf_counters.h:25]: This should return a unique_ptr<PerfCounters> and be used to directly initialize ss_perf in the constructor colon initialization list.

File: src/crimson/os/seastore/segment_cleaner.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:369]: It's mandatory, and SegmentCleaner doesn't own it.  Use a reference.

[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:369]: Same everywhere else in this PR.  SeaStore owns PerfService and is responsible for cleaning up the PerfCounters instance.  Every other constructor and reference to the PerfCounters reference (sc_perf and perf here) should be by reference and should be initialized directly in the constructor colon initializer list.

File: src/crimson/os/seastore/cache.cc (hunk 5 above)
Comment (tchaikov) [reply src/crimson/os/seastore/cache.cc:26]: have we looked into seastar's metric subsystem? is it a better alternative in comparison to `PerfCounters`?

Comment (liu-chunmei) [reply src/crimson/os/seastore/cache.cc:26]: > c_perf appears unused?

not yet.  currently no statistic information in your list need be collected from cache.  In case we need it later.  I can remove it in this PR if you think so.

Comment (athanatos) [reply src/crimson/os/seastore/cache.cc:26]: We can add it once there's a user.  Remove it for now.

File: src/crimson/os/seastore/seastore_perf_counters.h (hunk 8 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore_perf_counters.h:25]: I think all parts in seastore use one perfcounter,  so it build once and save to PerfService's ss_perf and each part use reference to this perfcounter.  do you mean each part such as transaction_manager and segment_manager build their own perfcounter and add them to perfcounters_collection?  ( each part does same add_to_collection and remove_to_collection?)

Comment (athanatos) [reply src/crimson/os/seastore/seastore_perf_counters.h:25]: No, I mean the ss_perf member should be a unique_ptr as I mentioned above and you should invoke this function in PerfService() to initialize it.

File: src/crimson/os/seastore/cache.cc (hunk 5 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/cache.cc]: > have we looked into seastar's metric subsystem? is it a better alternative in comparison to `PerfCounters`?

yingxin seems try it. I have no idea about it now.

**Comment by liu-chunmei** [comment]
@athanatos updated according to comments.  please review again.

File: src/crimson/os/seastore/seastore.h [hunk 9]
```diff
@@ -210,6 +212,7 @@ class SeaStore final : public FuturizedStore {
   TransactionManagerRef transaction_manager;
   CollectionManagerRef collection_manager;
   OnodeManagerRef onode_manager;
+  PerfService &perf_service;
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore.h:215]: SeaStore needs to actually own this.  Pass it into the constructor by value and this should be

```
PerfService perf_service;
```

Comment (athanatos) [reply src/crimson/os/seastore/seastore.h:215]: I'd also be fine with constructing it in place here and not passing it into the constructor.

File: src/crimson/os/seastore/segment_cleaner.cc [hunk 10]
```diff
@@ -345,6 +345,7 @@ SegmentCleaner::gc_reclaim_space_ret SegmentCleaner::gc_reclaim_space()
 	      ).safe_then([this, &t] {
 		if (scan_cursor->is_complete()) {
 		  t->mark_segment_to_release(scan_cursor->get_offset().segment);
+                  sc_perf.inc(ss_sm_segments_released);
```
Comment (athanatos) [inline src/crimson/os/seastore/segment_cleaner.cc:348]: There's a corresponding bit in TransactionManager::submit_transaction_direct where we actually do the release -- I think ss_sm_segment_released should be there.

**Comment by athanatos** [comment]
SegmentManager, BlockSegmentManager, and EphemeralSegmentManager all have unused PerfCounters references.  Please only include members that are actually in use in this PR.

File: src/crimson/os/seastore/seastore.h (hunk 9 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.h:215]: @athanatos seastore constructor need TransactionManagerRef and TransactionManagerRef constructor need perfcounter reference.  If I move perf_service initialized in seastore, then TransactionManager should be also initialized in seastore.  and TransactionManager  need journal cache initialized first.   So which do you prefer? initialize all TransactionManager  and PerfService outside of seastore constructor or move them into seastore ?

Comment (athanatos) [reply src/crimson/os/seastore/seastore.h:215]: Ah.  In that case you'll have to pass it in by value/rval reference.  SeaStore still needs to own it just as it does transaction_manager, collection_manager, onode_manager etc.

File: src/crimson/os/seastore/segment_cleaner.cc (hunk 10 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/segment_cleaner.cc:348]: @athanatos since sc_perf is defined private, can't  be called in transaction manager function. so I move sc_perf.inc(ss_sm_segments_released)  to segment_cleaner->mark_segment_released  function. Is it OK?

**Comment by liu-chunmei** [comment]
@athanatos updated according your comments, please review again.

**Comment by athanatos** [comment]
Make ss_perf private and add a PerfCounters &get_counters() method to replace the  *(perf_service->ss_perf) instances.  LGTM otherwise.

**Comment by liu-chunmei** [comment]
@athanatos updated. I think I'd better rebase it after your PR #41129 merged in since there are some conflict with your PR.

**Comment by liu-chunmei** [comment]
@athanatos rebased, please review.

**Comment by liu-chunmei** [comment]
@athanatos @tchaikov do you get agreement to merge perfcounters? if so, I will rebase it.

**Comment by athanatos** [comment]
Yep!

**Comment by tchaikov** [comment]
@liu-chunmei yes, please rebase. i will try to come up the follow-up changes to use `seastar::metrics`.

**Comment by liu-chunmei** [comment]
@athanatos rebased.

**Review by athanatos — APPROVED** [review]

**Comment by tchaikov** [comment]
> @liu-chunmei yes, please rebase. i will try to come up with follow-up changes to use `seastar::metrics`.

#41352
